### ISSUE TITLE
Adapt to email?

### ISSUE THREAD

**Issue by agenbite** [qadzek/vim-md-link#1]
Hi! This is a brilliant idea. For those of us who write emails with nvim, it'd be great to have a similar thing which got urls out of the way and down to a References section. I have tried with some macros, but your solution is much more elegant. I wish I could program in order to extend it for email!! :)

**Comment by agenbite** [comment]
Thanks for the quick response! :)

In text-based emails [0], sometimes it's nice to have references listed at the end, but this is far from being a standard. Some usual html to text converters do it this way too [1].

In any case, the issue is that in order to do that manually, you need to go to the bottom of your message, manually add the link and then back to where you were to write the square-bracket reference to the link. Of course, if you add a link in the first paragraph, you'll be forced to manually update the numbers of all your links after that one.

I tried something very rudimentary to ease this a bit in emacs+evil, but it just took care of the traversing of the document, and not really cared about numbering or anything like that.

I hope this is clearer now. I tried looking at your code, but vimscript is a mystery to me. Thanks again!! :)


Links:
[0] https://useplaintext.email/
[1] https://manpages.debian.org/testing/python3-html2text/html2markdown.1.en.html#reference

**Comment by Konfekt** [comment]
For what it's worth

```vim
if !empty(findfile('ftplugin/markdown/vim-md-link.vim', &rtp))
    runtime! ftplugin/markdown/vim-md-link.vim
endif

if exists(':MdLinkConvertAll') != 2 | finish | endif

autocmd vimrcFileTypeMail BufWrite <buffer> MdLinkConvertAll
```

provides the `:MdLink...` commands in mail buffers. 
The only nitpick would be that the reference section could be placed before the signature. 

I am using my fork of  [vim-alias](https://github.com/Konfekt/vim-alias) to abbreviate these commands

```
if exists(':Alias') != 2 | finish | endif

Alias -buffer lcs MdLinkConvertSingle
Alias -buffer lca MdLinkConvertAll
Alias -buffer lo  MdLinkOpen
Alias -buffer lp  MdLinkPeek
Alias -buffer lj  MdLinkJump
Alias -buffer ldu MdLinkDeleteUnneededRefs
```

but most will opt for mappings.

**Comment by Konfekt** [comment]
A possible adaption of :MdLinkConvertAll to Mails leaving the signature at the end:

```vim
command! -buffer MailLinkConvertAll call mdlink#ConvertAllLinks() |
            \ call MoveSignatureToEnd()

" The rest of the file needs to be :sourced only once per session.
if !exists('s:loaded_functions') || &cp
    let s:loaded_functions = 1
    function! MoveSignatureToEnd()
        " Find the line number of the signature line
        let dash_start = search('^-- ', 'W')
        " If not found, exit the function
        if dash_start == 0
            return
        endif

        " Find the line number of the line reading ## Links
        let links_start = search('^## Links$', 'W')
        " If not found, exit the function
        if links_start == 0
            return
        endif

        " Calculate the range of lines to move
        let move_start = dash_start
        let move_end = links_start - 1

        call append(line('$'), '')
        " If the range is valid, move the lines
        if move_start < move_end " && dash_start < links_start
            " Move lines after # Links to before --
            execute move_start . ',' . move_end . 'move $'
        endif
    endfunction
endif
```

**Comment by Konfekt** [comment]
Since in Mails, (https) links are often pasted, without proper markdown syntax `[...](...)`, how about numbering those and including them in the list of links?

**Comment by Konfekt** [comment]
> inline bar http://... and reference bar [3] "mail" links aren't supported at the moment.

It was meant as a substitute to merely pasting a link without naming it, as often happens in mails.
For the time being,

```vim
command! -buffer MailLinkConvertAll call
            \ ConvertLinks |
            \ mdlink#ConvertAllLinks() |
            \ call MoveSignatureToEnd()

command! -buffer -range=% -bar ConvertLinks silent keeppatterns <line1>,<line2>substitute#\v%(\[[^\]]+\]\(\s*)@<!\zs<https?://([^.]+)\S+#[\1](\0)#giep
```

that replaces the mere link by its name to the first domain might suffice

**Comment by agenbite** [comment]
Thanks, @Konfekt and @qadzek!! I'll (try to) find time to test the proposed solutions! :)

**Comment by Konfekt** [comment]
@agenbite It's now been somewhat officialized in the [wiki](https://github.com/qadzek/vim-md-link/wiki/Email)

**Comment by Konfekt** [comment]
There are two regexes, the first matches after the header lines, the second one before the signature or your link section heading. The syntax is that of `:help :range-pattern`.
Yes, maybe, but then I also have to check for the signature.
It would nice if your range function takes as optional parameters beginning and end markers, say.

**Comment by agenbite** [comment]
Wow, thank you so much, folks. This solution is just perfect for my needs!!! <3<3<3

**Comment by agenbite** [comment]
Wow, this is so great!! Thank you so much!! <3
